plugins/SpectrumAnalyzer/SaProcessor.h

```diff
@@ -0,0 +1,121 @@
+/* SaProcessor.h - declaration of SaProcessor class.
+ *
+ * Copyright (c) 2019 Martin Pavelek <he29/dot/HS/at/gmail/dot/com>
+ *
+ * Based partially on Eq plugin code,
+ * Copyright (c) 2014 David French <dave/dot/french3/at/googlemail/dot/com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef SAPROCESSOR_H
+#define SAPROCESSOR_H
+
+#include <QImage>
+#include <QMutex>
+#include <vector>
+
+#include "fft_helpers.h"
+#include "SaControls.h"
+
+
+//! Receives audio data, runs FFT analysis and stores the result.
+class SaProcessor
+{
+public:
+	explicit SaProcessor(SaControls *controls);
+	virtual ~SaProcessor();
+
+	void analyse(sampleFrame *in_buffer, const fpp_t frame_count);
+
+	// inform processor if any processing is actually required
+	void setSpectrumActive(bool active);
+	void setWaterfallActive(bool active);
+
+	// configuration is taken from models in SaControls; some changes require
+	// an exlicit update request (reallocation and window rebuild)
+	void reallocateBuffers();
+	void rebuildWindow();
+	void clear();
+
+	// information about results and unit conversion helpers
+	float binToFreq(int index);
+	float binBandwidth();
+
+	float freqToXPixel(float frequency, int width);
+	float xPixelToFreq(float x, int width);
+
+	float ampToYPixel(float amplitude, int height);
+	float yPixelToAmp(float y, int height);
+
+	int getSampleRate() const;
+	float getFreqRangeMin(bool linear = false);
+	float getFreqRangeMax();
+	float getAmpRangeMin(bool linear = false);
+	float getAmpRangeMax();
+
+	// data access lock must be acquired by any friendly class that touches
+	// the results, mainly to prevent unexpected mid-way reallocation
+	QMutex m_dataAccess;
+
+private:
+	SaControls *m_controls;
+
+	// currently valid configuration
+	const unsigned int m_zeroPadFactor = 2;	// use n-steps bigger FFT for given block
+	unsigned int m_inBlockSize;				// size of input data block
+	unsigned int m_fftBlockSize;			// size of padded block for FFT processing
```

Comment:
I'm not sure if there is a better name for it. It is size of the block that is going to be processed, so, well, it is named blockSize ^^; The second one could be also accurately named "transform size". And it applies to both time and frequency domain, in theory (because of the symmetry). In practice, binCount() is of course used to get the number of "useful" frequency domain samples. I will add "time domain" to the comments to make it more explicit.
(And as you probably figured out already, inBlockSize and fftBlockSize are different due to zero padding.)

Reply:
Oh, well, you are right, it's the same block size for frequency and time domain! In this case, ignore my comment, please. :+1:

Reply:
(or add to the comment that it's valid for both domains, if you think that makes sense)

Reply:
Saying that it applies to both time and frequency domain may be misleading without context, so I just tweaked the comments, adding that xxBlockSize is for time domain, binCount() is for frequency domain. And binCount() in .cpp has comment that explains why is it lower.